File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        close_to_open = {
           '}' : '{',
           ')' : '(',
           ']' : '['
        }

        for char in s:
            if char in close_to_open:
                if stack and stack[-1] == close_to_open[char]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(char)

        return len(stack) == 0                    
        

File: 0020-valid-parentheses/0020-valid-parentheses.dart
class Solution {
  bool isValid(String s) {
   List<String> stack = [];

    for(String bracket in s.split("")) {
      if(bracket == '(' || bracket == '[' || bracket == '{') {
        stack.add(bracket);
    }
    else {
      if(stack.isEmpty) return false;
      if(bracket == ')' && stack.last != '(') return false;
      if(bracket == '}' && stack.last != '{') return false;
      if(bracket == ']' && stack.last != '[') return false;
      
      stack.removeLast();
    }
  }
  return stack.isEmpty;
}
}
    